app/src/main/java/com/duckduckgo/app/trackerdetection/model/TdsTracker.kt

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2019 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.trackerdetection.model
+
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+import androidx.room.TypeConverter
+import com.duckduckgo.app.di.JsonModule
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.Types
+
+@Entity(tableName = "tds_tracker")
+data class TdsTracker(
+    @PrimaryKey val domain: String,
+    val defaultAction: Action,
+    val ownerName: String,
+    val categories: List<String>,
+    val rules: List<Rule>
+)
+
+enum class Action {
+    BLOCK,
+    IGNORE;
+}
+
+class Rule(
+    val rule: String,
+    val action: Action?,
+    val exceptions: RuleExceptions?
+)
+
+class RuleExceptions(
+    val domains: List<String>?,
+    val types: List<String>?
+)
+
+class ActionTypeConverter {
+
+    @TypeConverter
+    fun toAction(value: String): Action {
+        return Action.valueOf(value)
+    }
+
+    @TypeConverter
+    fun fromAction(value: Action): String {
+        return value.name
+    }
+}
+
+class RuleTypeConverter {
+
+    @TypeConverter
+    fun toRules(value: String): List<Rule> {
+        return Adapters.ruleListAdapter.fromJson(value)
+    }
+
+    @TypeConverter
+    fun fromRules(value: List<Rule>): String {
+        return Adapters.ruleListAdapter.toJson(value)
+    }
+}
+
+class CategoriesTypeConverter {
+
+    @TypeConverter
+    fun toCategories(value: String): List<String> {
+        return Adapters.stringListAdapter.fromJson(value)
+    }
+
+    @TypeConverter
+    fun fromCategories(value: List<String>): String {
+        return Adapters.stringListAdapter.toJson(value)
+    }
+}
+
+class Adapters {
+    companion object {
+        private val moshi = JsonModule().moshi()
```

Comment:
This is a bit icky - let me know if you have any ideas for how we can use DI to get a moshi instance into the converters.

Reply:
My suggestion here isn't much of an improvement, but here goes (based on https://stackoverflow.com/a/55600994/1654145)

1. Add a field to `AppDatabase`'s companion object, for Moshi
1. Pass the moshi instance through DI to `DatabaseModule` when creating the `AppDatabase` instance
1. Grab `moshi` instance from `AppDatabase.moshi`

I would say it's still not great, but it is a mild improvement over manually instantiation of `JsonModule`. what do you think?

Reply:
@Module(includes = [DaoModule::class])
    class DatabaseModule {

    @Provides
    @Singleton
    fun provideDatabase(context: Context, moshi: Moshi): AppDatabase {
        val db = Room.databaseBuilder(context, AppDatabase::class.java, "app.db")
            .addMigrations(*AppDatabase.ALL_MIGRATIONS.toTypedArray())
            .build()

        AppDatabase.moshi = moshi

        return db
    }



    abstract class AppDatabase : RoomDatabase() {

        ...

        companion object {

            lateinit var moshi: Moshi
        }
}

Reply:
Nice, I've merged but will add that to a tidy up PR after our conversation tomorrow.